cli/cook/subcommands/ssh.py

```diff
@@ -0,0 +1,112 @@
+    """Attempts to ssh (using os.execlp) to the Mesos agent corresponding to the given instance."""
+    hostname = instance['hostname']
+    if 'output_url' in instance:
+        output_url = instance['output_url']
+        url = urlparse(output_url)
+        netloc = url.netloc
+    else:
+        logging.info('assuming default agent port of 5051')
+        netloc = f'{hostname}:5051'
+
+    resp = http.__get(f'http://{netloc}/state')
+    if resp.status_code != 200:
+        logging.error(f'mesos agent returned status code {resp.status_code} and body {resp.text}')
+        raise Exception('Encountered error when querying Mesos agent for its work directory.')
+
+    resp_json = resp.json()
+    flags = resp_json['flags']
+    if 'work_dir' not in flags:
+        logging.error(f'work_dir was not present in mesos agent state flags: {flags}')
+        raise Exception('Mesos agent state did not include its work directory.')
+
+    print_info(f'Executing ssh to {colors.bold(hostname)}.')
+    agent_work_dir = flags['work_dir']
+    directory = sandbox_directory(agent_work_dir, instance, job)
+    file = os.environ.get('SSH', 'ssh')
+    logging.info(f'using ssh file: {file}')
+    os.execlp(file, 'ssh', '-t', hostname, f'cd {directory} ; bash')
+
+
+def ssh_to_job(job):
+    """
+    Tries ssh_to_instance for each instance from most to least recent, until either
+    ssh can be executed for an instance, or all instances have been tried and failed.
+    """
+    instances = job['instances']
+    if len(instances) == 0:
+        raise Exception('This job currently has no instances.')
+
+    for instance in sorted(instances, key=itemgetter('start_time'), reverse=True):
+        try:
+            print_info(f'Attempting ssh for job instance {colors.bold(instance["task_id"])}...')
+            ssh_to_instance(instance, job)
+        except Exception as e:
+            print_info(str(e))
+    else:
+        raise Exception(f'Unable to ssh for any instance of this job.')
+
+
+def ssh(clusters, args):
+    """Attempts to ssh (using os.execlp) to the Mesos agent corresponding to the given job or instance uuid."""
+    uuids = strip_all(args.get('uuid'))
+    if len(uuids) > 1:
+        # argparse should prevent this, but we'll be defensive anyway
+        raise Exception(f'You can only provide a single uuid.')
+
+    query_result = query(clusters, uuids)
+    num_results = query_result['count']
+
+    if num_results == 0:
+        print_no_data(clusters)
+        return 1
+
+    if num_results > 1:
+        # This is unlikely to happen in the wild, but it could.
+        # A couple of examples of how this could happen:
+        # - same uuid on an instance and a job (not necessarily the parent job)
+        # - same uuid on a job in cluster x as another job in cluster y
+        print_info('There is more than one match for the given uuid.')
+        return 1
+
+    for cluster_name, entities in query_result['clusters'].items():
```

Comment:
Is it possible for there to be multiple clusters?  Using `for` implies that it's possible.

Reply:
This command wants to resolve a uuid to a unique something.  If future commands wants the same semantics, it might be useful to have a function that works something like this:

<pre>
query_unique(uuid) -> {'type': 'job', 'data': {jobdata}}
                    | {'type': 'instance', 'data': {instance-data}}
                    | {'type': 'group', 'data': {group-data}}</pre>

That way this and other functions that want one thing don't have to do the same processing.

Reply:
Good call. Extracted a function `query_unique`. For now, it doesn't support a group uuid, because that's the same pattern we want for ssh, fuse, ls, tail